contracts/libs/data-structures/IncrementalMerkleTree.sol

```diff
@@ -278,25 +360,53 @@ library IncrementalMerkleTree {
         return tree.leavesCount;
     }
 
-    function _getZeroHashes(uint256 height_) private pure returns (bytes32[] memory) {
+    function _getZeroHashes(
+        IMT storage tree,
+        uint256 height_
+    ) private view returns (bytes32[] memory) {
         bytes32[] memory zeroHashes_ = new bytes32[](height_);
 
-        zeroHashes_[0] = ZERO_HASH;
+        zeroHashes_[0] = _getZeroHash(tree);
 
         for (uint256 i = 1; i < height_; ++i) {
-            bytes32 result;
             bytes32 prevHash_ = zeroHashes_[i - 1];
 
-            assembly {
-                mstore(0, prevHash_)
-                mstore(32, prevHash_)
+            zeroHashes_[i] = _hash2Fn(tree, prevHash_, prevHash_);
+        }
 
-                result := keccak256(0, 64)
-            }
+        return zeroHashes_;
+    }
 
-            zeroHashes_[i] = result;
+    function _getZeroHash(IMT storage tree) private view returns (bytes32) {
+        return _hash1Fn(tree, bytes32(0));
+    }
+
+    function _hash1Fn(IMT storage tree, bytes32 a) private view returns (bytes32 result) {
+        if (tree.isHashFnSet) {
+            return tree.hash1Fn(a);
         }
 
-        return zeroHashes_;
+        assembly {
+            mstore(0, a)
+
+            result := keccak256(0, 32)
+        }
+    }
+
+    function _hash2Fn(
+        IMT storage tree,
+        bytes32 a,
+        bytes32 b
+    ) private view returns (bytes32 result) {
+        if (tree.isHashFnSet) {
```

Comment:
It's more optimal to have a getter function that will return a hash function before the loops in  `_root`, `_add`, `_getZeroHashes` 

`function(bytes32, bytes32) view returns (bytes32) hash2_ = _getHash2(tree)`